### PR TITLE
Drop support for PHP 5.x, add PHP 7.0 features

### PR DESCRIPTION
**Pull request by chrisminett**
Drop support for PHP 5.x

Supporting >= PHP 7.0 allows use of (some) type declarations, and an upgrade for PHPUnit. It will require a major version increment.

I considered skipping straight to PHP 7.1, as 7.0 is no longer *actively supported*, but left it in as it's still receiving security updates.

The upgrade for PHPUnit broke some of the tests that were relying on a mock created from `StreamInterface` as it was trying to use a PHP global constant as if it were in the interface's namespace. Seemed simple enough to just use Guzzle's PSR7 `stream_for()` instead.

I've done this so that a new Factory I'm looking to add can be written for PHP 7.

### CODE REVIEW

File: src/Csv.php [hunk 1]
```diff
@@ -48,33 +50,24 @@ class Csv implements \Iterator, \Countable
...
+    public function getMaxColumns(): int
     {
         return $this->maxColumns;
     }
 
-    /**
-     * @param int $maxColumns
-     */
-    public function setMaxColumns($maxColumns)
+    public function setMaxColumns(int $maxColumns)
```
Comment (awhale) [inline src/Csv.php]: The mxaColumns parameter is being cast to int later in this method

**Comment by awhale** [comment]
Unless there is a reason not to, I would jump to 7.1 so that you can make use of the void return type to further define the method signatures.

**Comment by awhale** [comment]
You have removed the doc blocks for some but not all methods where the doc block definition doesn't add anything to the method signature.

**Comment by awhale** [comment]
Changing the PHPUnit assertions to be static is unusual and adds no value - so I wouldn't do that.

**Comment by chrisminett** [comment]
More docblocks can be removed if moving to php71 and using the `void` return type, so I'll make that change. Otherwise they were only left where they provide extra human description.

Re: PHPUnit statics. it's not unusual, it's correct. The methods are static, so should not be called dynamically.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 - CHANGELOG.md
+- PHP 7 type declarations for method parameters and return values
+### Removed
+- Drop support for PHP 5.x and PHP 7.1 as they are no longer
```
Comment (Pudge601) [inline CHANGELOG.md]: This is dropping support for PHP 7.0, no?

Comment (chrisminett) [reply CHANGELOG.md]: 👍 Fixed

**Review by Pudge601 — APPROVED** [review]

**Comment by awhale** [comment]
Static methods can also be called from object context using $this, and this is also correct.

In the case of PHPUnit, calling the assert methods from object context is the approach taken in the official documentation, as well as the best practice that the community have adopted in every open source project I have encountered. Any developer looking at the tests would consider the style unusual, which is why I don't think the change should be made without a stronger reason.

**Review by awhale — CHANGES_REQUESTED** [review]

**Comment by chrisminett** [comment]
I'll undo the `static::assert*` changes.

[PHPUnit documentation](https://phpunit.readthedocs.io/en/7.1/assertions.html#appendixes-assertions) says that either way is acceptable.

My sole reason for the change is that it's unusual to call static methods dynamically (although valid, unlike the other way around) and so my IDE reports warns on it.

File: src/Csv.php (hunk 1 above)
Comment (chrisminett) [reply src/Csv.php]: fixed

**Comment by awhale** [comment]
Cool, thanks. Lets chat next week about the statics separate to this PR, there's a really good PHPUnit extension in PHPStorm that I think is enabled by default now that might relieve most of your IDE warning pains - I was showing it to Pudge about a month ago

**Pull request merged**